### ISSUE TITLE
Question about all of the loss

### ISSUE THREAD

**Issue by zfs1212** [fungtion/DSN#5]
Thank you pytorch code. I want to ask some questions. First, I think the loss functions of " similarity" should be  achieved by "DANN" or "MMD" ,but you use Crossentropyloss. Second, loss = loss_classification + loss_diff + loss_similarity + loss_recon, however you are achieve loss = loss_classification + loss_diff + loss_recon1 + loss_recon2. I don't know if I got it wrong, Please point out my fault.THX

**Comment by fungtion** [comment]
First, my similarity function  is achieved by DANN, a batch of domain label can be written as a binary vector, e.g. [1, 1, 0, 1, 1] , and their groudtruth domain label is also can  be written as a binary vector, e.g. [1,1,1,1,1], so I use CrossEntropyLoss to compare them. 
Second, I also add the loss_similarity in my loss functions after **active_domain_loss_step**, see https://github.com/fungtion/DSN/blob/master/train.py/#L178

**Comment by zfs1212** [comment]
Thank you for your reply immediately.I don't know why the result of run the code is wrong--ACC is 0.000.The result is follow:
epoch: 0, accuracy of the mnist dataset: 0.000000
epoch: 0, accuracy of the mnist_m dataset: 0.000000
source_classification: 0.002207, source_dann: 0.000000, source_diff: 0.000000, source_mse: 0.003233, source_simse: 0.000000, target_dann: 0.000000, target_diff: 0.000000, target_mse: 0.002850, target_simse: 0.000348
epoch: 1, accuracy of the mnist dataset: 0.000000
epoch: 1, accuracy of the mnist_m dataset: 0.000000
source_classification: 0.001927, source_dann: 0.000000, source_diff: 0.000000, source_mse: 0.002884, source_simse: 0.000000, target_dann: 0.000000, target_diff: 0.000000, target_mse: 0.001150, target_simse: 0.000034
epoch: 2, accuracy of the mnist dataset: 0.000000
epoch: 2, accuracy of the mnist_m dataset: 0.000000
source_classification: 0.004587, source_dann: 0.000000, source_diff: 0.000000, source_mse: 0.003012, source_simse: 0.000000, target_dann: 0.000000, target_diff: 0.000000, target_mse: 0.001950, target_simse: 0.000086
epoch: 3, accuracy of the mnist dataset: 0.000000
epoch: 3, accuracy of the mnist_m dataset: 0.000000
......
What do you think caused it, please help me. THX

**Comment by fungtion** [comment]
I think it's the problem of the pytorch version, I tested it on 0.3.0, you can check it or tell me if you find any solution

**Comment by zfs1212** [comment]
Thank you very much for your kindness, your python version is python=2.7 and torch is 0.3.0 ?

**Comment by fungtion** [comment]
Yes

**Comment by JiaoJinyang** [comment]
Thanks for your code. But I have a problem. From the original paper, the loss function should be constructed by source data and target data, then the overall loss is trained. However, in you code, the target data and source data are trained separately. I don't konw it is right? Could you please explain this problem?

**Comment by fungtion** [comment]
Indeed, in each epoch, I trained the target data the first, and the source data the second, since when applying cross-entropy loss as similatiry in shared encoder, one of two inputs to the `torch.nn.CrossEntropyLoss` should not be the variable in the graph , and the other one should be the variable in the graph. For this reason, I must keep either output of the shared encoder of source data or target data invarible, when training  the other data.

**Comment by JiaoJinyang** [comment]
> Indeed, in each epoch, I trained the target data the first, and the source data the second, since when applying cross-entropy loss as similatiry in shared encoder, one of two inputs to the `torch.nn.CrossEntropyLoss` should not be the variable in the graph , and the other one should be the variable in the graph. For this reason, I must keep either output of the shared encoder of source data or target data invarible, when training the other data.

Thank you very much, but I think the total_loss should contain target_dann, target_diff, target_simse, source_dann, source_class, source_diff and source simse, then using the oprimizer to minimize this total_loss overall, not optimize target loss and source loss respectively. I don't know whether they are equivalent.
In addition, according the paper, the mse should substract the second term to get simse, but your code is add of two terms. Could you explain it? Thank you again!

**Comment by fungtion** [comment]
I think both training method can obtain similar results, you can try to rewrite the cross entropy loss and test it. For the second problem, in my experiment, it showed worse results than adding two items, so I changed it to this version.

**Comment by yingbaihu** [comment]
> Thank you very much for your kindness, your python version is python=2.7 and torch is 0.3.0 ?

I have the same problem with u, have u solved this problem? or just change the version of torch?

**Comment by lolinkun** [comment]
@fungtion Hi, thanks for your code. I am so interested in this paper, however, I am unable to reproduce your results.
My experiment environment is following your suggestion, I am not sure if my setting of the dataset is right?  
The dataset folder contains mnist folder and mnist_m folder. Mnist_m folder contains images as the dataset, however, mnist folder contains .gz files rather than images. Since I try to embedded images into mnist folder, however, my code is always not working. So I wrote 'download=True,' in the dataset.MNIST module of train.py. I am not sure if my implementation right or wrong?

**Comment by fungtion** [comment]
@lolinkun mnist file can be extracted using `tar -zvxf minst.tar.gz`. If you set  "download=True", you have to modify the path of mnist in the code.

**Comment by lolinkun** [comment]
@fungtion  Thanks for your reply. I can run the code now, however, I am suffering the accuracy almost 0 situations. I strictly follow your environment instruction as: python 2.7  torch==0.4.0, however, it seems not work.

**Comment by fungtion** [comment]
@lolinkun try to modify https://github.com/fungtion/DSN/blob/master/test.py#L144 to `accu = n_correct.data.numpy() * 1.0 / n_total`

**Comment by lolinkun** [comment]
> @lolinkun try to modify https://github.com/fungtion/DSN/blob/master/test.py#L144 to `accu = n_correct.data.numpy() * 1.0 / n_total`

Great！ You kindly replied my question, which really worked!

**Comment by lolinkun** [comment]
@fungtion 
Great！ You kindly replied my question, which really worked!
